Adapt gardener-dashboard component

grolu opened the pull request:
**How to categorize this PR?**

/area usability
/kind enhancement

**What this PR does / why we need it**:
This PR adapts the Gardener Operator logic for the Dashboard component to incorporate the latest improvements made in gardener/dashboard

**Which issue(s) this PR fixes**:
Fixes #10116

**Special notes for your reviewer**:
/cc @petersutter @holgerkoser

**Release note**:
<!--
Write your release note:
1. Enter your release note in the below block.
2. If no release note is required, just write "NONE" within the block.

Format of block header: <category> <target_group>
Possible values:
- category:       breaking|noteworthy|feature|bugfix|doc|other
- target_group:   user|operator|developer|dependency
-->
```other operator
NONE
```

grolu committed:
Added support for Dashboard session secret rotation

grolu committed:
do not create lease object for github webhook

grolu committed:
remove get secret permission

grolu committed:
remove ExperimentalUseWatchCacheForListShoots feature flag

grolu committed:
Fixed session secret vailidity

petersutter commented on pkg/component/gardener/dashboard/dashboard_test.go:
```diff
@@ -988,6 +994,38 @@ frontend:
+						Data: map[string][]byte{
+							"password": []byte("________________________________"),
+							"username": []byte("admin"),
+							"auth":     []byte("admin:$2y$05$VV/caJeJ0XEza7sc5hHib.uppkej805AYCGAKbSCbZwPz6INJy07G"),
```
use different fake password for previousSessionSecret

ScheererJ commented on pkg/component/gardener/dashboard/dashboard.go:
```diff
@@ -138,12 +136,18 @@ func (g *gardenerDashboard) Deploy(ctx context.Context) error {
+	sessionSecretPrevious, previousExists := g.getPreviousSessionSecret()
+	if previousExists {
+		sessionSecretPreviousName = sessionSecretPrevious.Name
+	}
```
`sessionSecretPrevious...` is a bit hard to read/comprehend. Would it be an option to rename to `oldSessionSecret` or something similar, which has a better reading flow?

grolu replied:
I understand, and we can rename it if you'd like. However, the current name aligns with the name in the dashboard configuration. That said, I'm happy to rename it if you prefer.

ScheererJ replied:
Personally, I prefer readable variable names. However, consistency is also important. Improving the variable name locally might reduce consistency globally due to the dashboard configuration naming.
I leave it up to you to decide how to resolve this conflict.

grolu committed:
PR Feedback
- use different fake password for previousSessionSecret
-  updated description to indicate that the clientSecret is optional

rfranzke commented on pkg/component/gardener/dashboard/secrets.go:
```diff
@@ -19,11 +20,14 @@ func (g *gardenerDashboard) newVirtualGardenAccessSecret() *gardenerutils.Access
+	}, secretsmanager.Rotate(secretsmanager.KeepOld), secretsmanager.Validity(30*24*time.Hour), secretsmanager.IgnoreOldSecretsAfter(24*time.Hour))
+}
+
+func (g *gardenerDashboard) getPreviousSessionSecret() (*corev1.Secret, bool) {
```
This function seems pretty useless - delete it and inline the call.

rfranzke commented on pkg/component/gardener/dashboard/dashboard.go:
```diff
@@ -138,12 +136,18 @@ func (g *gardenerDashboard) Deploy(ctx context.Context) error {
 
+	var sessionSecretPreviousName string
+	sessionSecretPrevious, previousExists := g.getPreviousSessionSecret()
+	if previousExists {
```
```suggestion
	if sessionSecretPrevious, ok := g.getPreviousSessionSecret(); ok {
```

rfranzke replied:
But why we need `sessionSecretPreviousName` at all? Isn't `sessionSecretPrevious` enough to pass to the `g.deployment()` call? Currently, you are checking there if the passed string (name) is non-empty. Instead, you could also simply pass the entire secret and check if it is non-nil.

grolu replied:
The reason I did this was because the deploy function parameters were all names and I wanted to align to that. However it indeed makes the code easier when I pass `sessionSecretPrevious` to `g.deployment()` so I changed it.

rfranzke commented on pkg/component/gardener/dashboard/deployment.go:
```diff
@@ -41,23 +42,29 @@ const (
-	volumeNameOIDC         = "gardener-dashboard-oidc"
-	volumeNameGithub       = "gardener-dashboard-github"
+	volumeSubPathSession           = "sessionSecret"
+	volumeMountPathSession         = "/etc/gardener-dashboard/secrets/session/sessionSecret"
```
Let's avoid this mixedCase in volume mounts. Use `-` instead
```suggestion
	volumeMountPathSession         = "/etc/gardener-dashboard/secrets/session/session-secret"
```

or even

```suggestion
	volumeMountPathSession         = "/etc/gardener-dashboard/secrets/session/secret"
```

grolu replied:
If we want to change this we need to adapt the implementation in g/dashboard first and make a release there.

rfranzke commented on pkg/component/gardener/dashboard/deployment.go:
```diff
@@ -41,23 +42,29 @@ const (
+	volumeSubPathSession           = "sessionSecret"
+	volumeMountPathSession         = "/etc/gardener-dashboard/secrets/session/sessionSecret"
+	volumeSubPathSessionPrevious   = "sessionSecretPrevious"
+	volumeMountPathSessionPrevious = "/etc/gardener-dashboard/secrets/session/sessionSecretPrevious"
```
same here

grolu committed:
PR Feedback #2
- Removed getPreviousSessionSecret function
- Pass previous session secret to deploy function

rfranzke reviewed (commented):
/lgtm
